README.markdown
= RSpec-Rails Extensions

Some prettifications for RSpec Rails.  

== Matchers

Some helpful matchers for testing HTTP statuses beyond the default be_success (200 OK)

    # looks for a 201 Created (useful for API CREATE calls)
    response.should be_successfully_created 
    
    # expects a Location field with the given URL (useful for API CREATE calls)
    response.should point_to(url) 
    
    # looks for a 422 response (invalid object)
    response.should be_unprocessable
    
    # looks for a 404
    response.should be_not_found
    
    # looks for a 401
    response.should be_unauthorised
    
    # looks for a 500
    response.should be_an_error
    
== Specifying your controllers

One of the issues with writing standard controller specifications, using RSpec's inbuilt mock objects is that the "flow" of your specs feels wrong.  

    it "should display a page for a given thingy" do
      @thingy = mock_model Thingy
      Thingy.should_receive(:find).with('1').and_return(@thingy)
      
      get :show, :id => '1'
      
      response.should be_success
      response.should render_template('thingies/show')
    end
    
In other words you are setting expectations before you are doing the work - technically correct but it reads wrongly.  

These extra helpers let you write specifications in a more english-like manner.  

    it "should create a new thingy" do
      @thingy = mock_model Thingy
      
      on_getting :show, :id => '1' do
        Thingy.should_receive(:find).with('1').and_return(@thingy)
      end

      response.should be_success
      response.should render_template('thingies/show')
    end
    
Isn't that lovely?

CHANGELOG
v1.8.1  Bugfix in ComparableArray
v1.8.   Added ComparableArray
v1.7.1. Added be_not_acceptable
v1.7.   Added use_layout matcher
v1.6.3. Changed the output when status/locations don't match
v1.6.2. Changed how we match the status header
v1.6.1. Added associated_model builder
v1.6.0. Pushed a version that actually works (oops, sorry about that)
v1.005. Added invalid_model
v1.004. Added `rake build`
v1.003. Added mocks - prepare_for_errors_on
v1.002. Bugfix
v1.001. Bugfix
v1.000. Initial release

spec/lib/rspec_rails_extensions/comparable_array_spec.rb
require File.expand_path(File.dirname(__FILE__) + "/../../../lib/rspec_rails_extensions/comparable_array.rb")

describe ComparableArray do
  describe "#=~" do
    before { @a = ComparableArray.new([:foo, :bar]) }
    it "should match identical array" do
      (@a =~ [:foo, :bar]).should == true
    end
    it "should match array with same elements in different order" do
      (@a =~ [:bar, :foo]).should == true
    end
    it "should not match array with different number of elements" do
      (@a =~ [:foo, :bar, :sed]).should == false
    end
    it "should not match array with same number of different elements" do
      (@a =~ [:foo, :sed]).should == false
    end
  end
end

describe Array do
  describe "~" do
    it "should wrap self in ComparableArray" do
      (~[]).should be_a_kind_of(ComparableArray)
      (~[:foo]).should be_a_kind_of(ComparableArray)
      (~[:foo]).array.should == [:foo]
    end
  end
end

describe Object do
  describe "~" do
    it "should return self" do
      o = Object.new
      (~o).should == o
    end
  end
end


lib/mocks.rb
module Spec
  module Rails
    module Mocks
      # Set up the given mock model in preparation for an ActiveRecord::RecordInvalid call
      # For example: 
      #   @user = mock_model User
      #   prepare_for_errors_on @user
      #   @user.should_receive(:save!).and_raise(ActiveRecord::RecordInvalid.new(@user))
      # If you tried this without the call to prepare_for_errors_on, the RecordInvalid exception would not be raised.  
      def prepare_for_errors_on model
        errors = mock 'errors'
        errors.should_receive(:full_messages).and_return([])
        model.should_receive(:errors).and_return(errors)
      end
    
      # Shortcut to create a mock model that is prepared for a ActiveRecord::RecordInvalid.new(@model) call
      # Equivalent to: 
      #   @model = mock_model Model, :some => :fields
      #   prepare_for_errors_on @model
      def invalid_model class_name, stubs = {}
        model = mock_model class_name, stubs
        prepare_for_errors_on model
        return model
      end
      
      # Shortcut to create a model that will be used in an association
      # Example: 
      #   @model = MasterModel.create!
      #   @mock = associated_model ChildModel, :some => :value
      #   @model.children << @mock
      def associated_model class_name, stubs = {}
        mock_model class_name, stubs.merge(:[]= => true, :save => true)
      end
    end
  end
end


lib/controller_example_group.rb
module Spec
  module Rails
    module Example
      class ControllerExampleGroup < FunctionalExampleGroup
        # Perform a pseudo HTTP GET, performing any code in the block beforehand.  This is to allow you to set up expectations on your mocks whilst keeping the code readable.   
        # For example: 
        #   on_getting :show, :id => '25' do
        #     @person = mock_model Person
        #     Person.should_receive(:find).with('25').and_return(@person)
        #   end
        # If you want an XMLHTTPRequest pass in :as_xhr => true in the parameters
        # (don't worry - it won't be passed to your controller)
        # For example:
        #   on_getting :show, :id => '25', :as_xhr => true do
        #     whatever
        #   end
        def on_getting page, parameters = {}
          yield if block_given?
          is_xhr_request = parameters.delete(:as_xhr)
          if is_xhr_request
            xhr :get, page, parameters
          else
            get page, parameters
          end
        end

        # Perform a pseudo HTTP POST, performing any code in the block beforehand.  This is to allow you to set up expectations on your mocks whilst keeping the code readable.   
        # As with on_getting, use :as_xhr => true to make this an XMLHTTPRequest
        def on_posting_to page, parameters = {}
          yield if block_given?
          is_xhr_request = parameters.delete(:as_xhr)
          if is_xhr_request
            xhr :post, page, parameters
          else
            post page, parameters
          end
        end
        
        # Perform a pseudo HTTP PUT, performing any code in the block beforehand.  This is to allow you to set up expectations on your mocks whilst keeping the code readable.   
        # As with on_getting, use :as_xhr => true to make this an XMLHTTPRequest
        def on_putting_to page, parameters = {}
          yield if block_given?
          is_xhr_request = parameters.delete(:as_xhr)
          if is_xhr_request
            xhr :put, page, parameters
          else
            put page, parameters
          end
        end
        
        # Perform a pseudo HTTP DELETE, performing any code in the block beforehand.  This is to allow you to set up expectations on your mocks whilst keeping the code readable.   
        # As with on_getting, use :as_xhr => true to make this an XMLHTTPRequest
        def on_deleting_from page, parameters = {}
          yield if block_given?
          is_xhr_request = parameters.delete(:as_xhr)
          if is_xhr_request
            xhr :delete, page, parameters
          else
            delete page, parameters
          end
        end
      end
    end
  end
end

lib/responses.rb
=begin rdoc
Module with a set of matchers for examining ActionController::Response objects
=end
module ResponseMatchers
  # test for a 201 response
  def be_successfully_created 
    ResponseStatusMatcher.new "201 Created"
  end
  
  # test for a 422 response
  def be_unprocessable
    ResponseStatusMatcher.new "422 Unprocessable Entity"
  end 
  
  # test for a 404 response
  def be_not_found
    ResponseStatusMatcher.new "404 Not Found"
  end
  
  # test for a 401 response
  def be_unauthorised
    ResponseStatusMatcher.new "401 Unauthorized"
  end
  alias :be_unauthorized :be_unauthorised
  
  # test for a 406 response
  def be_not_acceptable
    ResponseStatusMatcher.new "406 Not Acceptable"
  end
  
  # test for a 500 internal server error 
  def be_an_error
    ResponseStatusMatcher.new "500 Internal Error"
  end
  
  # test that the location points to a given URI
  def point_to url
    ResponseLocationMatcher.new url
  end
  
  # Response matcher that examines ActionController::Response headers for the required status code
  class ResponseStatusMatcher
    # Set up this matcher as required
    def initialize status_code
      @status_code = status_code
    end

    # Does the given target object match the required status code?
    def matches? target
      (@target_status = target.status) == @status_code
    end
    
    # What do we tell the user when it fails?
    def failure_message
      a = <<-EOF
      expected the location header to be #{@status_code.inspect}
                          instead it was #{@target_status.inspect}
      EOF
    end
    
    # What do we tell the user when it shouldn't fail but does
    def negative_failure_message
      "expected the response to be different to #{@status_code.inspect}"
    end
  end
  
  # Response matcher that examines ActionController::Response headers for the given location code
  class ResponseLocationMatcher
    def initialize url
      @url = url
    end
    
    # Does the given target object match the required location?
    def matches? target
      (@target_url = target.headers['Location']) == @url
    end
    
    # What do we tell the user when it fails?
    def failure_message
      a = <<-EOF
      expected the location header to be #{@target_url.inspect}
                          instead it was #{@url.inspect}
      EOF
    end
    
    # What do we tell the user when it shouldn't fail but does
    def negative_failure_message
      "expected the location header to be different to #{@url.inspect}"
    end
  end

end